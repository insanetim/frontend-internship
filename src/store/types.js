export const LOADING_ON = 'app/LOADING_ON'
export const LOADING_OFF = 'app/LOADING_OFF'
export const SHOW_NOTIFICATION = 'app/SHOW_NOTIFICATION'
export const SHOW_MODAL = 'app/SHOW_MODAL'
export const HIDE_MODAL = 'app/HIDE_MODAL'
export const LOG_IN = 'session/LOG_IN'
export const LOG_IN_CANCEL = 'session/LOG_IN_CANCEL'
export const LOG_OUT = 'session/LOG_OUT'
export const SET_SESSION = 'session/SET_SESSION'
export const DELETE_SESSION = 'session/DELETE_SESSION'
export const REQUEST_ACCOUNT = 'account/REQUEST_ACCOUNT'
export const SET_ACCOUNT = 'account/SET_ACCOUNT'
export const REQUEST_TRENDING = 'trending/REQUEST_TRENDING'
export const SET_TRENDING = 'trending/SET_TRENDING'
export const REQUEST_SEARCH = 'search/REQUEST_SEARCH'
export const SET_SEARCH = 'search/SET_SEARCH'
export const SET_SEARCH_QUERY = 'search/SET_SEARCH_QUERY'
export const CLEAR_SEARCH = 'search/CLEAR_SEARCH'
export const REQUEST_LISTS = 'lists/REQUEST_LISTS'
export const SET_LISTS = 'lists/SET_LISTS'
export const REQUEST_LIST = 'lists/REQUEST_LIST'
export const SET_LIST = 'lists/SET_LIST'
export const CREATE_LIST = 'lists/CREATE_LIST'
export const DELETE_LIST = 'lists/DELETE_LIST'
export const ADD_TO_LIST = 'lists/ADD_TO_LIST'
export const REMOVE_FROM_LIST = 'lists/REMOVE_FROM_LIST'
export const REQUEST_WATCHLIST = 'watchlist/REQUEST_WATCHLIST'
export const SET_WATCHLIST = 'watchlist/SET_WATCHLIST'
export const REQUEST_FAVORITES = 'favorites/REQUEST_FAVORITES'
export const SET_FAVORITES = 'favorites/SET_FAVORITES'
export const REQUEST_MOVIE = 'movie/REQUEST_MOVIE'
export const SET_MOVIE = 'movie/SET_MOVIE'
export const SET_MOVIE_IN_FAVORITE = 'movie/SET_MOVIE_IN_FAVORITE'
export const CHANGE_MOVIE_IN_FAVORITE = 'movie/CHANGE_MOVIE_IN_FAVORITE'
export const SET_MOVIE_IN_WATCHLIST = 'movie/SET_MOVIE_IN_WATCHLIST'
export const CHANGE_MOVIE_IN_WATCHLIST = 'movie/CHANGE_MOVIE_IN_WATCHLIST'
