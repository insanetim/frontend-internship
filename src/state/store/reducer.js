import { combineReducers } from 'redux'

import appReducer from '../app'
import sessionReducer from '../session'
import dashboardReducer from '../dashboard'
import listsReducer from '../lists'
import watchlistReducer from '../watchlist'
import favoritesReducer from '../favorites'
import movieReducer from '../movie'

const rootReducer = combineReducers({
  app: appReducer,
  session: sessionReducer,
  dashboard: dashboardReducer,
  lists: listsReducer,
  watchlist: watchlistReducer,
  favorites: favoritesReducer,
  movie: movieReducer
})

export default rootReducer
