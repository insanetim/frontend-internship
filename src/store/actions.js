const { LOG_IN, LOADING_ON, LOADING_OFF, SHOW_NOTIFICATION, SET_SESSION, LOG_OUT } = require('./types')

export const loadingOn = () => ({
  type: LOADING_ON
})

export const loadingOff = () => ({
  type: LOADING_OFF
})

export const login = (payload, cb) => ({
  type: LOG_IN,
  payload,
  cb
})

export const logout = cb => ({
  type: LOG_OUT,
  cb
})

export const setSession = payload => ({
  type: SET_SESSION,
  payload
})

export const showNotification = payload => ({
  type: SHOW_NOTIFICATION,
  payload
})
