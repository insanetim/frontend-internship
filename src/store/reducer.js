import * as types from './types'

const initialState = {
  sessionId: null,
  account: {},
  trending: {},
  searchQuery: null,
  searchResult: {},
  lists: {},
  watchlist: {},
  favorites: {},
  modalType: null,
  modalProps: {},
  loading: false
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case types.LOADING_ON:
      return { ...state, loading: true }
    case types.LOADING_OFF:
      return { ...state, loading: false }
    case types.SHOW_MODAL:
      return { ...state, modalType: action.payload.modalType, modalProps: action.payload.modalProps }
    case types.HIDE_MODAL:
      return { ...state, modalProps: { open: false } }
    case types.SET_SESSION:
      return { ...state, sessionId: action.payload }
    case types.DELETE_SESSION:
      return { ...state, sessionId: null }
    case types.SET_ACCOUNT:
      return { ...state, account: action.payload }
    case types.SET_TRENDING:
      return { ...state, trending: action.payload }
    case types.SET_SEARCH_QUERY:
      return { ...state, searchQuery: action.payload }
    case types.SET_SEARCH_RESULT:
      return { ...state, searchResult: action.payload }
    case types.CLEAR_SEARCH:
      return { ...state, searchQuery: null, searchResult: {} }
    case types.SET_LISTS:
      return { ...state, lists: action.payload }
    case types.SET_WATCHLIST:
      return { ...state, watchlist: action.payload }
    case types.REQUEST_FAVORITES:
      return { ...state, favorites: action.payload }
    case types.SET_FAVORITES:
      return { ...state, favorites: action.payload }
    default:
      return state
  }
}
