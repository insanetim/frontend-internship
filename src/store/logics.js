import { createLogic } from 'redux-logic'
import { notification as AntdNotification } from 'antd'
import * as Cookies from 'js-cookie'
import { merge } from 'lodash'

import * as endpoints from 'src/constants/endpoints'
import * as types from './types'
import * as actions from './actions'

const notification = createLogic({
  type: types.SHOW_NOTIFICATION,
  latest: true,
  process({
    action: {
      payload: { type, message }
    }
  }) {
    AntdNotification[type]({ message })
  }
})

const showModal = createLogic({
  type: types.SHOW_MODAL,
  latest: true,
  transform({ action }, next) {
    const newAction = merge(action, {
      payload: {
        modalProps: {
          open: true
        }
      }
    })
    next(newAction)
  }
})

const login = createLogic({
  type: types.LOG_IN,
  cancelType: types.LOG_IN_CANCEL,
  latest: true,
  async process({ httpClient, action: { payload, cb } }, dispatch, done) {
    dispatch(actions.loadingOn())
    const { data: requestTokenData } = await httpClient.get(endpoints.createRequestToken)
    const { data: sessionToken } = await httpClient
      .post(endpoints.createSessionWithLogin, {
        request_token: requestTokenData.request_token,
        ...payload
      })
      .catch(error => {
        dispatch(actions.showNotification({ type: 'error', message: error.response.data.status_message }))
        dispatch(actions.loadingOff())
        dispatch({ type: types.LOG_IN_CANCEL })
        done()
      })
    const { data } = await httpClient.post(endpoints.createSession, {
      request_token: sessionToken.request_token
    })
    Cookies.set('session_id', data.session_id)
    dispatch(actions.loadingOff())
    cb()
    done()
  }
})

const logout = createLogic({
  type: types.LOG_OUT,
  latest: true,
  async process({ httpClient, getState, action: { cb } }, dispatch, done) {
    const { sessionId } = getState()
    const { data } = await httpClient.delete(endpoints.deleteSession, {
      data: { session_id: sessionId }
    })
    if (data.success) {
      Cookies.remove('session_id')
      dispatch(actions.deleteSession())
      cb()
    }
    done()
  }
})

const requestAccount = createLogic({
  type: types.REQUEST_ACCOUNT,
  latest: true,
  async process({ httpClient, getState }, dispatch, done) {
    const { sessionId } = getState()
    const { data } = await httpClient.get(endpoints.getDetails, {
      params: {
        session_id: sessionId
      }
    })
    dispatch(actions.setAccount(data))
    done()
  }
})

const requestTrending = createLogic({
  type: types.REQUEST_TRENDING,
  latest: true,
  async process({ httpClient, action: { payload: page = 1 } }, dispatch, done) {
    const { data } = await httpClient.get(endpoints.getTrending, {
      params: {
        page
      }
    })
    dispatch(actions.setTrending(data))
    done()
  }
})

const requestSearch = createLogic({
  type: types.REQUEST_SEARCH,
  latest: true,
  async process(
    {
      httpClient,
      action: {
        payload: { query, page = 1 }
      }
    },
    dispatch,
    done
  ) {
    dispatch(actions.setSearchQuery(query))
    const { data } = await httpClient.get(endpoints.searchMovies, {
      params: { query, page }
    })
    dispatch(actions.setSearchResult(data))
    done()
  }
})

const requestLists = createLogic({
  type: types.REQUEST_LISTS,
  latest: true,
  async process({ httpClient, getState, action: { payload: page = 1 } }, dispatch, done) {
    const {
      sessionId,
      account: { id }
    } = getState()
    const { data } = await httpClient.get(endpoints.getCreatedLists(id), {
      params: { session_id: sessionId, page }
    })
    dispatch(actions.setLists(data))
    done()
  }
})

const createList = createLogic({
  type: types.CREATE_LIST,
  latest: true,
  async process({ httpClient, getState, action: { payload } }, dispatch, done) {
    const { sessionId } = getState()
    const { data } = await httpClient.post(
      endpoints.createList,
      { ...payload },
      {
        params: {
          session_id: sessionId
        }
      }
    )
    if (data.success) {
      dispatch(actions.requestLists())
    }
    done()
  }
})

const deleteList = createLogic({
  type: types.DELETE_LIST,
  latest: true,
  async process({ httpClient, getState, action: { payload: id } }, dispatch, done) {
    const { sessionId } = getState()
    await httpClient
      .delete(endpoints.deleteList(id), {
        params: { session_id: sessionId }
      })
      .finally(() => dispatch(actions.requestLists()))
    done()
  }
})

const requestWatchlist = createLogic({
  type: types.REQUEST_WATCHLIST,
  latest: true,
  async process({ httpClient, getState, action: { payload: page = 1 } }, dispatch, done) {
    const {
      sessionId,
      account: { id }
    } = getState()
    const { data } = await httpClient.get(endpoints.getWatchlist(id), {
      params: { session_id: sessionId, page }
    })
    dispatch(actions.setWatchlist(data))
    done()
  }
})

const requestFavorites = createLogic({
  type: types.REQUEST_FAVORITES,
  latest: true,
  async process({ httpClient, getState, action: { payload: page = 1 } }, dispatch, done) {
    const {
      sessionId,
      account: { id }
    } = getState()
    const { data } = await httpClient.get(endpoints.getFavorites(id), {
      params: { session_id: sessionId, page }
    })
    dispatch(actions.setFavorites(data))
    done()
  }
})

export default [
  notification,
  showModal,
  login,
  logout,
  requestAccount,
  requestTrending,
  requestSearch,
  requestLists,
  createList,
  deleteList,
  requestWatchlist,
  requestFavorites
]
