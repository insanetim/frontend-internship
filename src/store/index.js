import { configureStore } from '@reduxjs/toolkit'
import { createLogicMiddleware } from 'redux-logic'
import httpClient from 'src/api/httpClient'

import logics from './logics'
import reducer from './reducer'

const deps = { httpClient }
const logicMiddleware = createLogicMiddleware(logics, deps)

const store = configureStore({
  reducer,
  middleware: [logicMiddleware],
  devTools: process.env.NODE_ENV !== 'production'
})

export default store
