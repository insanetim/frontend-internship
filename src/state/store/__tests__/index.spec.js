import store from '../index'

it('matches snapshot', () => {
  expect(store).toMatchSnapshot()
})
