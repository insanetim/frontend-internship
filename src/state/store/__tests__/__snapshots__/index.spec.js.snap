// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
{
  "@@observable": [Function],
  "dispatch": [Function],
  "getState": [Function],
  "replaceReducer": [Function],
  "subscribe": [Function],
}
`;
